### ISSUE TITLE
Permission handling allows users to lose write access.

### ISSUE THREAD

**Issue by jSML4ThWwBID69YC** [moosefs/moosefs#354]
## Have you read through available documentation and open Github issues?

Yes

Note: https://moosefs.com/wp-content/uploads/2018/08/MooseFS-Hardware-Guide-v.0.9.pdf is a 404 error. 

## Is this a BUG report, FEATURE request, or a QUESTION? Who is the indended audience?

BUG report

#### Your moosefs version and its origin (moosefs.com, packaged by distro, built from source, ...).

All componets built from the FreeBSD ports tree. 

moosefs3-master-3.0.112_1
moosefs3-cgiserv-3.0.112_1
moosefs3-master-3.0.112_1
moosefs3-chunkserver-3.0.112_1
moosefs3-client-3.0.112_1
fusefs-libs3-3.9.1

#### Operating system (distribution) and kernel version.

FreeBSD 12.1p3

#### Hardware / network configuration, and underlying filesystems on master, chunkservers, and clients.

MooseFS master server: 1x
CPU: Intel(R) Xeon(R) CPU E5-2637 v2 @ 3.50GHz
RAM: 256GB 
HW DISK: 4x 1.2TB in raidz1

Moosefs chunk servers: 5x identical 
CPU: Intel(R) Xeon(R) CPU E5-2407 0 @ 2.20GHz
RAM: 32GB
HW DISK: 4x 8TB in raidz1. 
Moose disk is a single folder using ZFS reservation and quotas mounted at /storage/chunk
DATA: storage/chunk 889G 18.1T 889G  /storage/chunk

MooseFS client servers: 2x identical
CPU: Intel(R) Xeon(R) CPU E5-2667 0 @ 2.90GHz
RAM: 192GB
HW DISK: 4x 1.2TB configured as raidz1

#### Network
All servers are plugged in twice with 10GB uplinks. The uplinks are configured as LACP across two Cisco switches in a stack. The network cards are Dell branded Intel cards using the 'ix' driver. 

#### How much data is tracked by moosefs master (order of magnitude)?

 - All fs objects: 10700038
 - Total space: 95 TiB
 - Free space: 91 TiB
 - RAM used: 3.9 GiB
 - last metadata save duration: ~3.6s

## Describe the problem you observed.

Permissions are handled differently on MooseFS than on FreeBSD. In this case, files and folders can become non writable, even when owned by the user. This differs from normal file permissions where a user can always make a file writable again.  

Here's a test case. Run the following as a normal unprivileged user on ZFS. 

1: touch test.txt

2: ls -l test.txt
-rw-r--r--  1 test  test  0 Apr  7 22:14 test.txt

3: chmod u-w test.txt

4: ls -l test.txt
-r--r--r--  1 test  test      0 Apr  7 22:14 test.txt

5: chmod u+w test.txt

6: ls -l test.txt
-rw-r--r--  1 test  test  0 Apr  7 22:14 test.txt

You can see that the user test can make the file test.txt non writable, and then writable again as the ownership of the file has not changed. 

Now run the same test again on MooseFS using the same unprivileged user.

1: touch test.txt

2: ls -l test.txt
-rw-r--r--  1 test  test  0 Apr  7 22:17 test.txt

3: chmod u-w test.txt

4: ls -l test.txt
-r--r--r--    1 test  test        0 Apr  7 22:17 test.txt

5: chmod u+w test.txt
chmod: test.txt: Permission denied

6: ls -l test.txt
-r--r--r--  1 test  test  0 Apr  7 22:17 test.txt

In step five, the user is no longer able to make the file writable. At this point, only the root user can make the file writable again. 

Expected behavior: Files should be able to be made writable by the file owner on MooseFS, as that is how it works on a local file system.

**Comment by jSML4ThWwBID69YC** [comment]
Note, this happens regardless of if mfsmkdircopysgid=1 is set or not. My initial test had it set, but followup tests show the same issue without it.

**Comment by jSML4ThWwBID69YC** [comment]
I'm not sure if this helps, but it does not happen with read or execute permissions. For example, 
chmod u-r test.txt followed by chmod u+r test.txt work as expected. It's only the write permission that's affected.

**Comment by pkonopelko** [comment]
> Note: https://moosefs.com/wp-content/uploads/2018/08/MooseFS-Hardware-Guide-v.0.9.pdf is a 404 error.

Hardware Guide URL redirect is fixed now.

**Comment by acid-maker** [comment]
MooseFS handles this correctly. On Linux it works fine:

```
$ touch test.txt
$ ls -l test.txt
-rw-r--r-- 1 core core 0 Apr  8 08:19 test.txt
$ chmod u-w test.txt
$ ls -l test.txt
-r--r--r-- 1 core core 0 Apr  8 08:19 test.txt
$ chmod u+w test.txt
$ ls -l test.txt
-rw-r--r-- 1 core core 0 Apr  8 08:19 test.txt
```

Likely FreeBSD does something extra during this chmod and it causes this error - I'll check it and let you know - hopefully it can be worked around in our code.

**Comment by acid-maker** [comment]
FreeBSD kernel before setattr sends access with 'W_OK' flag set (which is incorrect - we do not need write access in case of setattr). Because all other calls have their own protections likely the only way is to make stupid workaround in access (in case of FreeBSD always return ok). Nevertheless we will submit the bug in FreeBSD.

**Comment by jSML4ThWwBID69YC** [comment]
@chogata 

Was the kernel bug reported upstream?

**Comment by chogata** [comment]
Yes, here: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=245689

**Comment by jSML4ThWwBID69YC** [comment]
Awesome. It looks like the fix made it into FreeBSD 12.2+ Is the MooseFS workaround still in place?

**Comment by chogata** [comment]
Yes, it's still in place. We have no way of knowing if MooseFS runs on a kernel with this fix, so we keep such workarounds until the "bad" version of system is really obsolete.
If this becomes a big issue for users, we can add a config option.
